core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4

```diff
@@ -841,6 +856,121 @@ pathSpecification
     : pathElement (',' pathElement)*
     ;
 
+functionSpecification
+    : FUNCTION qualifiedName parameterDeclarationList
+        returnsClause routineCharacteristic*
+        (returnStatement | compoundStatement)
+    ;
+
+parameterDeclarationList
+    : '(' (parameterDeclaration (',' parameterDeclaration)*)? ')'
+    ;
+
+parameterDeclaration
+    : identifier? type
+    ;
+
+returnsClause
+    : RETURNS type
+    ;
+
+routineCharacteristic
+    : LANGUAGE identifier               #languageCharacteristic
+    | NOT? DETERMINISTIC                #deterministicCharacteristic
+    | RETURNS NULL ON NULL INPUT        #returnsNullOnNullInputCharacteristic
+    | CALLED ON NULL INPUT              #calledOnNullInputCharacteristic
+    | SECURITY (DEFINER | INVOKER)      #securityCharacteristic
+    | COMMENT string                    #commentCharacteristic
+    ;
+
+routineStatement
+    : returnStatement
+    | assignmentStatement
+    | simpleCaseStatement
+    | searchedCaseStatement
+    | ifStatement
+    | iterateStatement
+    | leaveStatement
+    | compoundStatement
+    | loopStatement
+    | whileStatement
+    | repeatStatement
+    ;
+
+returnStatement
+    : RETURN valueExpression
+    ;
+
+assignmentStatement
+    : SET identifier EQ expression
```

Comment:
Standard SQL also allows:

```
SET (a, b, c, d) = (1, 2, 3, 4)
```

We can leave that for later, though.

Reply:
Yeah, I didn't implement row destructuring.